src/lib.rs

```diff
@@ -0,0 +1,780 @@
+    }
+
+    #[inline(always)]
+    pub fn kind(self) -> AllocationKind {
+        match (self.bits >> 29) & 3 {
+            0 => AllocationKind::None,
+            1 => AllocationKind::Reg,
+            2 => AllocationKind::Stack,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn index(self) -> usize {
+        (self.bits & ((1 << 29) - 1)) as usize
+    }
+
+    #[inline(always)]
+    pub fn as_reg(self) -> Option<PReg> {
+        if self.kind() == AllocationKind::Reg {
+            Some(PReg::from_index(self.index()))
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    pub fn as_stack(self) -> Option<SpillSlot> {
+        if self.kind() == AllocationKind::Stack {
+            Some(SpillSlot(self.index() as u32))
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    pub fn bits(self) -> u32 {
+        self.bits
+    }
+
+    #[inline(always)]
+    pub fn from_bits(bits: u32) -> Self {
+        Self { bits }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(u8)]
+pub enum AllocationKind {
+    None = 0,
+    Reg = 1,
+    Stack = 2,
+}
+
+impl Allocation {
+    #[inline(always)]
+    pub fn class(self) -> RegClass {
+        match self.kind() {
+            AllocationKind::None => panic!("Allocation::None has no class"),
+            AllocationKind::Reg => self.as_reg().unwrap().class(),
+            AllocationKind::Stack => self.as_stack().unwrap().class(),
+        }
+    }
+}
+
+/// A trait defined by the regalloc client to provide access to its
+/// machine-instruction / CFG representation.
+pub trait Function {
+    // -------------
+    // CFG traversal
+    // -------------
+
+    /// How many instructions are there?
+    fn insts(&self) -> usize;
+
+    /// How many blocks are there?
+    fn blocks(&self) -> usize;
+
+    /// Get the index of the entry block.
+    fn entry_block(&self) -> Block;
+
+    /// Provide the range of instruction indices contained in each block.
+    fn block_insns(&self, block: Block) -> InstRange;
+
+    /// Get CFG successors for a given block.
+    fn block_succs(&self, block: Block) -> &[Block];
+
+    /// Get the CFG predecessors for a given block.
+    fn block_preds(&self, block: Block) -> &[Block];
+
+    /// Get the block parameters for a given block.
+    fn block_params(&self, block: Block) -> &[VReg];
+
+    /// Determine whether an instruction is a call instruction. This is used
+    /// only for splitting heuristics.
+    fn is_call(&self, insn: Inst) -> bool;
+
+    /// Determine whether an instruction is a return instruction.
+    fn is_ret(&self, insn: Inst) -> bool;
+
+    /// Determine whether an instruction is the end-of-block
+    /// branch. If so, its operands *must* be the block parameters for
+    /// each of its block's `block_succs` successor blocks, in order.
+    fn is_branch(&self, insn: Inst) -> bool;
```

Comment:
Yes, that's true; however designing that restriction into the interface I think might create unnecessary pain for some producers/clients that want to avoid such optimizations, or strive for simplicity in general. It's easy enough to allow unnecessary blockparams so I don't see the harm in allowing this :-)